src/Main.elm

```diff
@@ -1,10 +1,418 @@
 module Main exposing (main)
 
-import Data.Audience
-import Data.AudienceFolder
-import Html exposing (Html)
+import Browser
+import Data.Api as Api
+import Data.Audience as Audience exposing (Audience, AudienceType)
+import Data.AudienceFolder exposing (AudienceFolder)
+import Data.Store as Store exposing (AudienceFolderID, AudienceLevel, Store)
+import Element exposing (Element, column, el, none, paragraph, row, text)
+import Element.Background
+import Element.Border
+import Element.Events
+import Element.Font
+import Element.Input exposing (button)
+import Html
+import Html.Attributes
+import Http
+import Task
 
 
-main : Html msg
+
+---------------
+-- M O D E L --
+---------------
+
+
+type alias InitialisingState =
+    { folders : Maybe (List AudienceFolder)
+    , audiences : Maybe (List Audience)
+    }
+
+
+type Location
+    = Root
+    | Folder AudienceFolderID
+
+
+type alias SucceedState =
+    { location : Location
+    , filterBy : AudienceType
+    }
+
+
+type Model
+    = Initialising InitialisingState
+    | Failed Http.Error
+    | Succeed Store SucceedState
+
+
+makeSelector : SucceedState -> Store.Selector
+makeSelector state =
+    case state.filterBy of
+        Audience.Authored ->
+            case state.location of
+                Root ->
+                    Store.OnlyAuthored
+
+                Folder folderID ->
+                    Store.OnlyAuthoredIn folderID
+
+        Audience.Shared ->
+            Store.OnlyShared
+
+        Audience.Curated ->
+            case state.location of
+                Root ->
+                    Store.OnlyCurated
+
+                Folder folderID ->
+                    Store.OnlyCuratedIn folderID
+
+
+turnInitialising : InitialisingState -> Model
+turnInitialising state =
+    case Maybe.map2 Store.init state.folders state.audiences of
+        Nothing ->
+            Initialising state
+
+        Just store ->
+            Succeed store (SucceedState Root Audience.Authored)
```

Comment:
there I think avoid pattern matching would be nicer.

```elm
Maybe.map2 Store.init state.folders state.audiences
  |> Maybe.map (Basics.Extra.flip Succeed (SucceedState Root Audience.Authored))
  |> Maybe.withDefault (Initialising state)
```

though readability with pattern matching is not worse, maybe even better if one preferse to think about branches rather than composition.

Reply:
I've gone thru the time when I tried to avoid any `case..of` but use some sort of flows and `map*` and `flip` and `curry` and `uncurry`... Then I realised that `case..of` is 10x times more clear and readable than deep nesting of parentheses. That's also what @rtfeldman told in the Elm Europe 2019 conference as well.

Using of `Basics.Extra.flip` adds one more extra knowledge which developers should know. It's easy but I'm 100% sure it doesn't help really much. It just making thinks more wrapped and requires more time to read, understand and change.

What will happen when `Succeed` get one more argument? Let's say it become `Succeed Store SucceedState Int`:

```elm
turnInitialising : Int -> InitialisingState -> Model
turnInitialising someInt state =
    Maybe.map2 Store.init state.folders state.audiences
        |> Maybe.map (flip Succeed (SucceedState Root Audience.Authored))
        |> Maybe.map ((|>) someInt)
        |> Maybe.withDefault (Initialising state)
```

or 

```elm
turnInitialising : Int -> InitialisingState -> Model
turnInitialising someInt state =
    Maybe.map2 Store.init state.folders state.audiences
        |> Maybe.map ((|>) someInt << flip Succeed (SucceedState Root Audience.Authored))
        |> Maybe.withDefault (Initialising state)
```

VS

```elm

turnInitialising : Int -> InitialisingState -> Model
turnInitialising someInt state =
    case Maybe.map2 Store.init state.folders state.audiences of
        Nothing ->
            Initialising state

        Just store ->
            Succeed store (SucceedState Root Audience.Authored) someInt
```

I bet the last piece of code is quite more readable and easy changeable.

Reply:
I know Richard's arguments and I don't say they have no point but there are also important contra arguments. And I don't mean the fact that pattern matching makes compiler slower, that's not even that big deal since 0.19.

Pattern matching is a good tool for dealing with branching. But in my opinion, branching should not be done everywhere in the code. More likely it's desirable to avoid branching in some parts of logic and push branching to a very low level and very high-level parts of control flow as much as possible. using pattern matching one can do almost anything in any of branches as long as the result type checks. The composition is much simpler to reason about as it must obey certain rules. Also writing code that can compose often leads to smaller and better self-contained building blocks. These composition rules are the primary tool we use in FP to reason about the program so it's required to be comfortable with them. Those patterns often let you define small functions that you can compose using the map and then build bigger blocks. With pattern matching this was of building larger structures from smaller ones is much more obfuscated.

The burden of choosing the right approach is still on a developer. But I think the argument of having to know less is something I can't buy into generally. If anything people should be encouraged to learn more.

Reply:
Agree, sometimes casing is not needed especially when you work with well known and final solutions, like `Maybe`, `Result`, etc. But in case when you use custom types in a way like here it means that the types will change often and so it's not a good place to try piping.